<h3 class="text-white">License and attribution</h3>
<p class="text-white mt-4">
    Data is available for academic and commercial use, under a <a href="https://creativecommons.org/licenses/by/4.0/">CC-BY-4.0</a> licence.
</p>
<p class="text-white">If you make use of a {{siteName}} prediction, please cite the following papers:</p>
<p class="text-white">
    <a href="https://doi.org/10.1038/s41592-022-01488-1">Mirdita M et al. ColabFold - Making Protein folding accessible to all. <em>Nature</em> (2022)</a><br/>
    <a href="https://www.biorxiv.org/content/10.1101/2021.10.04.463034v2">Evans R et al. Protein complex prediction with AlphaFold-Multimer. <em>bioRxiv</em> (2021)</a><br/>
    <a href="https://doi.org/10.1038/s41586-021-03819-2">Jumper, J et al. Highly accurate protein structure prediction with AlphaFold. <em>Nature</em> (2021)</a><br/>
    <a href="https://doi.org/10.1093/bioinformatics/bty1057">Mirdita M et al. MMseqs2 desktop and local web server app for fast, interactive sequence searches. <em>Bioinformatics</em> (2019)</a><br/>
    <a href="https://doi.org/10.1093/nar/gkw1081">Mirdita M et al. Uniclust databases of clustered and deeply annotated protein sequences and alignments. <em>Nucleic Acid Res.</em> (2017)</a><br/>
    <a href="https://doi.org/10.1093/nar/gkz1035">Mitchell AL et al. MGnify: the microbiome analysis resource in 2020. <em>Nucleic Acids Research</em> (2019)</a><br/>
    <a href="https://doi.org/10.1371/journal.pcbi.1005659">Eastman P et al. OpenMM 7: Rapid development of high performance algorithms for molecular dynamics. <em>PLOS Comput. Biol.</em> (2017)</a><br>
    <a href="#">Samuel Schwab, Aimee L. Boyle, and Remus T. Dame. Novel histones and histone variant families in prokaryotes. <em>bioRxiv</em> (2023)</a><br>
    <a href="#">Samuel Schwab, Tim Jansen, Michel J. Olsthoorn, and Remus T. Dame. ProHistoneDB: the prokaryotic histone database. <em>In preparation</em> (2024)</a>
</p>